managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -1793,6 +1830,34 @@ protected void internalAsyncMarkDelete(final PositionImpl newPosition, Map<Strin
     }
 
     void internalMarkDelete(final MarkDeleteEntry mdEntry) {
+        if (persistentMarkDeletePosition != null
+                && mdEntry.newPosition.compareTo(persistentMarkDeletePosition) < 0) {
+            if (log.isInfoEnabled()) {
+                log.info("Skipping updating mark delete position to {}. The persisted mark delete position {} "
+                        + "is latest.", mdEntry.newPosition, persistentMarkDeletePosition);
+            }
+            mdEntry.triggerComplete();
+            return;
+        }
+
+        PositionImpl inProgressLatest = INPROGRESS_MARKDELETE_PERSIST_POSITION_UPDATER.updateAndGet(this, current -> {
+            if (current != null && current.compareTo(mdEntry.newPosition) > 0) {
+                return current;
+            } else {
+                return mdEntry.newPosition;
+            }
+        });
+
+        // if there's a newer or equal mark delete update in progress, skip it.
+        if (inProgressLatest != mdEntry.newPosition) {
+            if (log.isInfoEnabled()) {
+                log.info("Skipping updating mark delete position to {}. The mark delete position update "
+                        + "in progress {} is latest.", mdEntry.newPosition, inProgressLatest);
+            }
+            mdEntry.triggerComplete();
```

Comment:
Is it correct to complete this `MarkDeleteEntry`? It seems to me that it's ambiguous at this point in the code because the current `inProgressLatest` might be in the process of persisting that position as mark deleted. I'd think we should comply complete that `MarkDeleteEntry` once the in progress mark delete is completed.

Reply:
You are right that it would seem like the correct solution. It would add more complexity to solve it in that way. 
I doubt that there would be an actual benefit. I'd rather solve this later in a way where concurrency is eliminated by using a single threaded model similar to an actor model.